hail/src/main/scala/is/hail/io/bgen/IndexBgen.scala

```diff
@@ -67,39 +72,42 @@ object IndexBgen {
 
     val sHadoopConfBc = hc.sc.broadcast(new SerializableHadoopConfiguration(hConf))
 
-    val rvds = headers.map { f =>
-      val partition = IndexBgenPartition(
-        f.path,
-        f.compressed,
-        skipInvalidLoci,
-        recoding,
-        f.dataStart,
-        f.fileByteSize,
-        0,
-        sHadoopConfBc)
-
-      val crvd = BgenRDD(hc.sc, Array(partition), settings, null)
-      OrderedRVD.coerce(typ, crvd)
-    }
-
     val rowType = typ.rowType
     val offsetIdx = rowType.fieldIdx("offset")
-    val (keyType, kf) = rowType.select(Array("locus", "alleles"))
+    val (keyType, _) = rowType.select(Array("locus", "alleles"))
 
     val attributes = Map("reference_genome" -> rg.orNull,
       "contig_recoding" -> recoding,
       "skip_invalid_loci" -> skipInvalidLoci)
 
-    val unionRVD = RVD.union(rvds)
-    assert(unionRVD.getNumPartitions == files.length)
-
-    unionRVD.toRows.foreachPartition({ it =>
-      val partIdx = TaskContext.get.partitionId()
-      using(new IndexWriter(sHadoopConfBc.value.value, indexFilePaths(partIdx), keyType, annotationType, attributes = attributes)) { iw =>
-        it.foreach { row =>
-          iw += (row.deleteField(offsetIdx), row.getLong(offsetIdx), Row())
-        }
-      }
-    })
+    implicit val ec = ExecutionContext.fromExecutor(Executors.newFixedThreadPool(headers.length))
+
+    val rvdFutures = headers.zipWithIndex.map { case (f, i) =>
+      Future({
+        val partition = IndexBgenPartition(
+          f.path,
+          f.compressed,
+          skipInvalidLoci,
+          recoding,
+          f.dataStart,
+          f.fileByteSize,
+          0,
+          sHadoopConfBc)
+
+        val crvd = BgenRDD(hc.sc, Array(partition), settings, null)
+        assert(crvd.getNumPartitions == 1)
+
+        OrderedRVD.coerce(typ, crvd).toRows.foreachPartition({ it =>
+          using(new IndexWriter(sHadoopConfBc.value.value, indexFilePaths(i), keyType, annotationType, attributes = attributes)) { iw =>
+            it.foreach { row =>
+              iw += (row.deleteField(offsetIdx), row.getLong(offsetIdx), Row())
+            }
+          }
+        })
+        info(s"Finished indexing file '${ f.path }'")
+      })
+    }
+
+    Await.result(Future.sequence(rvdFutures.toFastIndexedSeq), Duration.Inf)
```

Comment:
This works?  I didn't realize Spark could run multiple jobs simultaneously with a single Spark context.

I'm worried about this because I think we've assumed the driver is single-threaded.  I think instead of coerce it is better to mapPartitions after the union and sort within the foreachPartition.